scipy/optimize/optimize.py

```diff
@@ -3061,6 +2912,567 @@ def show_options(solver=None, method=None, disp=True):
+            return self.x.size
+        elif self.x0 is not None:
+            return self.x0.size
+        return RuntimeError("Cannot determine problem size at this time,"
+                            " perform at least one iteration")
+
+    def solve(self):
+        """
+        Run the solver through to completion.
+
+        Returns
+        -------
+        result: OptimizeResult
+            The optimization result represented as a `OptimizeResult` object.
+            Important attributes are: `x` the solution array, `success` a
+            Boolean flag indicating if the optimizer exited successfully and
+            `message` which describes the cause of the termination. See
+            `OptimizeResult` for a description of other attributes.
+        """
+        for x, fun in self:
+            self.x, self.fun = x, fun
+
+            if (self.converged() or self.nfev >= self.maxfun or
+                    self.nit >= self.maxiter):
+                break
+
+        self._finish_up()
+        return self.result
+
+    def converged(self):
+        """
+        The truth of whether the solver has converged.
+        """
+        raise NotImplementedError
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """
+        Advance the solver by a single iteration.
+        """
+        # Should be over-ridden by each class based solver.
+        # TODO, Stop if self.maxfun/self.maxiter reached?
+        raise NotImplementedError
+    # step is because PyTorch has a step method
+    next = step = __next__
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self._finish_up()
+
+    def _finish_up(self):
+        """
+        Each class based solver should override this to define end-of-run
+        actions. This includes setting the `message` status, `warn_flag`,
+        `success`, etc. It gets called at the end of the `solve` and `__call__`
+        methods, and during `__exit__` of the context manager.
+        """
+        pass
+
+    def show_options(self):
+        # each optimizer should return their own options text.
+        return ""
+
+
+class Function(object):
+    def __init__(self, func=None, args=(), kwargs=None, jac=None, hess=None):
+        self._func = func
+        self.args = args
+        self.kwargs = {}
+        if kwargs is not None:
+            self.kwargs.update(kwargs)
+        self._jac_fun = jac
+        self._hess_fun = hess
+
+        self._last_x = None
+        self._last_fun = None
+        self._last_jac = None
+
+        self.nfev = 0
+        self.njev = 0
+        self.nhev = 0
+
+    def _same_x(self, x):
+        return numpy.array_equal(x, self._last_x)
+
+    # TODO thought of using decorators to count calls of func, jac, hess,
+    # but overridden methods would probably forget to do so.
+    def __call__(self, x):
+        # do not override this method otherwise nfev may not be correct
+        self.nfev += 1
```

Comment:
Having state in the Function object is dangerous, because it is tempting to reuse it. Leaving aside the exact actual way to advance the iteration, you may want to do something like:

```
f = Function(my_func, jac=my_grad)

with NelderMead(f, x0, , maxfun=100) as opt:
    while opt.step() and not opt.converged():
        print('Still going...')

with LBFGSB(f, x0, maxfun=100) as opt:
    while opt.step() and not opt.converged():
        print('Still going...')

```

And the second time around you'd start counting function evaluations where the first optimization finished, which is probably not what you want to do...

Reply:
That's a very good point. Not sure how to proceed on this one. I tried using decorators but that won't work with an overridden method. Any suggestuona.?

Reply:
Not really... The obvious solution seems to be to keep all the state in one place. I'd say in the Optimizer? It would make sense especially if it operates as a context manager, but I'm not sure how that could work. One can also think of making Function a context manager, and have a factory function so that you could do:

```
f = Function(...)

with f.reset_all() as f_:
    # f_ is a copy of f with all evaluations reset
```

But frankly, it is a little cumbersome...

I need to find the time to review your code in depth and think this whole thing through, but I wanted to leave a note here so I didn't forget.

Reply:
But storing the number of calls to the function/jacobian/hessian in `Function` is natural and where I'd expect to find these values.

We could have the optimizer reset the state of the function when initialized.